cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -331,26 +344,75 @@ func doInstall(mst *initramfsMountsState, model *asserts.Model, sysSnaps map[sna
 		return fmt.Errorf("cannot use gadget: %v", err)
 	}
 
-	// TODO:COMPS take into account kernel-modules components, see
-	// DeviceManager,doSetupRunSystem and other parts of
-	// handlers_install.go.
+	// Get kernel-modules information to have them ready early on first boot
 
-	bootDevice := ""
-	kernelSnapInfo := &gadgetInstall.KernelSnapInfo{
-		Name:       kernelSnap.SnapName(),
-		MountPoint: kernelMountDir,
-		Revision:   kernelSnap.Revision,
-		// Should be true always anyway
-		IsCore: !model.Classic(),
-	}
-	switch model.Base() {
-	case "core20", "core22", "core22-desktop":
-		kernelSnapInfo.NeedsDriversTree = false
-	default:
-		kernelSnapInfo.NeedsDriversTree = true
+	kernCompsByName := make(map[string]*snap.Component)
+	for _, c := range kernelSnap.Components {
+		kernCompsByName[c.Name] = c
+	}
+
+	kernelSeed := sysSnaps[snap.TypeKernel]
+	kernCompsMntPts := make(map[string]string)
+	compSeedInfos := []install.CompSeedInfo{}
+	for _, sc := range kernelSeed.Components {
+		seedComp := sc
+		comp, ok := kernCompsByName[seedComp.CompSideInfo.Component.ComponentName]
+		if !ok {
+			return fmt.Errorf("component %s in seed but not defined by snap!",
+				seedComp.CompSideInfo.Component.ComponentName)
+		}
+		if comp.Type != snap.KernelModulesComponent {
+			continue
+		}
+
+		// Mount ephemerally the kernel-modules components
+		mntPt := filepath.Join(filepath.Join(boot.InitramfsRunMntDir, "snap-content",
+			seedComp.CompSideInfo.Component.String()))
+		if err := doSystemdMount(seedComp.Path, mntPt, &systemdMountOptions{
+			ReadOnly:  true,
+			Private:   true,
+			Ephemeral: true}); err != nil {
+			return err
```

Comment:
Do we need a test in this error case to ensure the unmounts are called correctly?

Reply:
I've added a test now